File: Multiverse.SimulationSandbox/SettlerScript.lua
function list_to_table(list)
    local table = {};
    local i
    for i = 0, list.Count - 1 do
        table[i + 1] = list[i]
    end
    return table
end

function filter_by_type(type, table)
    local i = 0
    local newtable = {}
    for _, unit in ipairs(table) do
        if unit.type == type then
            i = i + 1
            newtable[i] = unit
        end
    end
    return newtable
end

function first(table)
    if type(table) ~= 'table' then
        table = list_to_table(table)
    end

    for _, item in ipairs(table) do
        return item
    end
    return nil
end

function onevent(self, event)
    local forests = filter_by_type(Forest, list_to_table(scanning.scanAround(self)))
    local forest = first(forests)
    if forest then
        if forest.x == self.x and forest.y == self.y then
            local harvestresult = abilities.use(self, 'HarvestWood', forest)
            print('harvest wood: ' .. harvestresult.type)
        else
            local movementability = first(self.abilities:withType(AbilityTypeMovement))
            local moveresult = abilities.use(self, movementability.name, nil, forest.x, forest.y)
            print('move to forest: ' .. moveresult.type)
        end
    end
end

File: Multiverse.SimulationSandbox/SettlerScript.ts
/// <reference path="multiverse.d.ts" />

debugging.log('script compiles ok');

function onevent(self: UnitSelf, event: ScriptingEvent): string {
    //debugging.log('event ' + event.type + ' happened at ' + event.timestamp);
    //const forest = scanning.scanAround(self).filter(x => x.type === Forest).shift();
    //if (forest) {
    //    if (forest.x === self.x && forest.y === self.y) {
    //        const harvestresult = abilities.use(self, 'HarvestWood', forest);
    //        debugging.log('harvest wood: ' + harvestresult.type);
    //    } else {
    //        const movementability = self.abilities.withType(AbilityTypeMovement).shift();
    //        const moveresult = abilities.use(self, movementability.name, null, forest.x, forest.y);
    //        debugging.log('move to forest: ' + moveresult.type);
    //    }
    //}

    const forest = scanning.scanAround(self).filter(x => x.type === Forest).shift();
    const buildingSite = scanning.scanAround(self).filter(x => x.type === WarehouseBuildingSite).shift();
    const warehouse = scanning.scanAround(self).filter(x => x.type === Warehouse).shift();
    const t = event.timestamp;

    if (forest && warehouse) {
        abilities.use(self, 'HarvestWood', forest);
        abilities.transferResource(self, warehouse, Wood, 1000);
    }

    if (t === 1) {
        abilities.use(self, 'Move', null, forest.x, forest.y);
    }
    if (t === 2) {
        abilities.startBuildingSite(self, WarehouseBuildingSite);
    }
    if (t === 3) {
        abilities.use(self, 'HarvestWood', forest);
        abilities.use(self, 'HarvestWood', forest);
        abilities.use(self, 'HarvestWood', forest);
        abilities.transferResource(self, buildingSite, Wood, 1000);
    }
    if (t === 4) {
        abilities.use(self, 'HarvestWood', forest);
        abilities.use(self, 'HarvestWood', forest);
        abilities.use(self, 'HarvestWood', forest);
        abilities.transferResource(self, buildingSite, Wood, 1000);
    }
    if (t === 5) {
        abilities.build(self, buildingSite);
    }
    if (t === 6) {
        abilities.build(self, buildingSite);
    }

    return null;
}
